acp/acp.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+/*
+Package acp utilizes the sourcehub acp module to bring the functionality to defradb, this package also helps
+avoid the leakage of direct sourcehub references through out the code base, and eases in swapping
+between local embedded use case and a more global on sourcehub use case.
+*/
+
+package acp
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/logging"
+)
+
+var (
+	log = logging.MustNewLogger("acp")
+
+	NoACPModule = immutable.None[ACPModule]()
+)
+
+// ACPModule is the interface to all types of access control modules there might exist.
+//
+// An embedded local implementation might restrict the implementation to not make remote calls.
+//
+// A remote implementation might allow remote calls to access the acp module in the cloud.
+// Note: sourcehub acp module will likely need the remote implementation.
+type ACPModule interface {
+
+	// Start initializes the acp module, with persistence data stored under the given folder.
+	//
+	// If previously started in the same path, then recovers / reloads (unless used in-memory).
+	Start(context.Context, string) error
+
+	// Close closes the resources in use by the acp module.
+	Close() error
+
+	// AddPolicy attempts to add the given policy. Detects the format of the policy automatically
+	// by assuming YAML format if JSON validation fails. Upon success a policyID is returned,
+	// otherwise returns error.
+	//
+	// A policy can not be added without a valid creator.
+	AddPolicy(ctx context.Context, creator string, policy string) (string, error)
```

Comment:
suggestion: `creatorID`? (I'm guessing a bit, but I assume this is not a json object :)).

Reply:
I don't follow the suggestion.

If it is to rename the `creator`, it should be `creatorIdentity`. LMK if you like that and I can do that.

Reply:
> If it is to rename the creator

Yes :)

> it should be creatorIdentity

Why? ID is short for Identity - why the preference for the long variant?

Reply:
because the "DefraDB Identity" / "Sourcehub Identity"  (name isn't finalized yet) and "will be soon" made up of mutiple components (sourcehub address, pub-priv key pair, a nounce). The creator is a simple identity right now which is to be more accurate a cosmos address (which will soon migrate to the source address).

I guess a more accurate word here is `actorID` of type SourceHubAddress/CosmosAddress.

So how about `creatorAddress`?

Thinking of it more `creatorID` is fine too.

Reply:
- [ ] rename to `creatorID`